SPEC.md

```diff
@@ -0,0 +1,225 @@
+## Apollo Fetchers
+
+The purpose of apollo-fetcher is to define an extensible standard interface for fetching GraphQL results. To that end, apollo-fetcher describes an interface containing a single method that connects a GraphQL `Operation` (also called request), to an `Observable` which delivers the results of the operation. Operations that return a single result can easily be mapped to a promise, since the observable's `next` function is called only once. In the general case, Apollo Fetcher uses observables to support GraphQL subscriptions and live queries. The apollo-fetcher's `Observable` follows the ECMAScript [proposal](https://github.com/tc39/proposal-observable), adding a `status` function. A basic fetcher is visualized as follows:
+
+<p align="center">
+  <br>
+  <img src="images/apollo-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+Fetchers are designed for current Apollo Client use-cases and to provide room for extensions. In addition to supporting Apollo Client, Apollo Fetchers are designed to be used independently as a very simple GraphQL client. These are the use cases that fetchers should support:
+
+* Do a simple GraphQL query with it that resolves only once
+* Do a simple GraphQL mutation that resolves only once
+* Do a subscription
+* Do a GraphQL query that contains @defer, @stream or @live
+* Send extra metadata along with queries
+* Support middleware to modify headers
+* Support middleware to modify query, variables, operationName and metadata
+* Isomorphic: It works in the browser and on the server
+* Support afterware to modify response
+* Support HTTP  transport, websocket transport, hybrid transport
+* Support alternative transports
+* Support per request retries based on network status or response
+* Support HTTP transport batching
+* Support polling
+* Support batching
+* Request deduplication of on-the-wire requests
+* Support splitting: eg. send queries and mutations to different servers
+
+Since this list focuses on Apollo Client's needs, please open a Pull Request adding to this list as you think of your desired use cases.
+
+In order to support all these use cases with a simple fetcher design, we want to use a modular architecture that allows fetchers to delegate to other fetchers.
+
+## Example
+
+This repository will contain a family of canonical fetchers that provide example extensions. The goal of these canonical fetchers is to be self-contained without changing the context of each operation.
+
+The current sample is http-fetcher, which sends an operation over http to a server at a specific URI. The uri and fetch function are passed to the constructor. This Http-fetcher can be used as a stand-alone GraphQL client:
+
+<p align="center">
+  <br>
+  <img src="images/http-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+```js
+const httpFetcher = new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' });
```

Comment:
I kind of feel like fetchers could be functions instead of classes/instances, and that would be more aligned with the current JS trends.

Why not:

```js
const httpFetcher = createHTTPFetcher({ uri: '...' });
const observable = httpFetcher({ ... });
```